### ISSUE TITLE
Hercules 4.4.1 crashes after OSA failure

### ISSUE THREAD

**Issue by cfdonatucci** [SDL-Hercules-390/hyperion#489]
___
_**NOTE:** &nbsp;GitHub Issue [\#458 (Hercules crash after resume from suspend)](https://github.com/SDL-Hercules-390/hyperion/issues/458) is also closely related to this issue._
___

Hi guys,

I had to reinstall my Windows 10 after a problem. After reinstalling Hercules, I'm now having an odd problem not happening before, regarding my external connection to use other Windows applications on the same PC. I reinstalled all Hercules software.

I also ran TTTest64.exe successfully. I have DHCP default conf in my PC. I always used CTCI connection but fails with no error message. So I tried with OSA, and now I get a dump.

z/OS 2.4 starts ok and the OSA is installed and activated ok.

I can start TSO sessions from any PC in my net, I can use CICSPLEX very well using `cicsexpl`.
When I attempt to use a `zosexpl` session with RSED, the connection crashed and Hercules as well.

Dump available:

* [Hercules1.zip](https://github.com/SDL-Hercules-390/hyperion/files/8642470/Hercules1.zip)

I'd appreciate any help!

Regards,
Carlos

### OSA Adapter
```
0400.3    QETH  chpid F0 iface D8-5E-D3-81-FE-1D  ipaddr 192.168.1.115  netmask 255.255.0.0

; -------------------------------------------------------------------
; Device to support zPDT External   to z/OS Connections
; -------------------------------------------------------------------
  DEVICE PORTA MPCIPA
  LINK OSA1 IPAQENET  PORTA
  HOME 192.168.1.115  OSA1
; -------------------------------------------------------------------
; Routes to support zPDT External   to z/OS Connections
; -------------------------------------------------------------------
  BEGINRoutes
  ; Destination        SubnetMask    FirstHop      LinkName    Size
  ROUTE 192.168.1.0    255.255.255.0    =          OSA1   MTU 1500
  ; Destination                      First Hop     LinkName    Size
  ROUTE DEFAULT                      192.168.1.1   OSA1   MTU 1500
  ENDRoutes
; -------------------------------------------------------------------
; Start  to support zPDT External   to z/OS Connections
; -------------------------------------------------------------------
  START PORTA
```
```
 BROWSE    ADCD.Z24B.VTAMLST(OSATRL2) - 01.01       Line 0000000000
 Command ===>                                                  Scrol
********************************* Top of Data **********************
OSATRL1 VBUILD TYPE=TRL
OSATRL1E TRLE LNCTL=MPC,READ=(0400),WRITE=(0401),DATAPATH=(0402),
               PORTNAME=PORTA,
               MPCLEVEL=QDIO
OSATRL2E TRLE LNCTL=MPC,READ=(0404),WRITE=(0405),DATAPATH=(0406),
               PORTNAME=PORTB,
               MPCLEVEL=QDIO
```
### Hercules log
```
12:32:36.422 00000788 HHC03800I 0:0401 QETH: Adapter mode set to Layer 3
12:32:36.422 00000788 HHC04100I TunTap64.dll version ** UNPAID TRIAL COPY **  3.7.0.5409 initiated
12:32:36.423 00000788 HHC00901I 0:0401 QETH: Interface tun0, type TUN opened
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using MAC address 02:00:5e:a3:be:84
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using IP address 192.168.1.115
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using subnet mask 255.255.0.0
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using MTU 1500
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using drive MAC address 96:7a:59:e5:d2:bf
12:32:36.492 00000788 HHC03997I 0:0401 QETH: tun0: using drive IP address fe80::967a:59ff:fee5:d2bf
12:32:36.499 00000788 HHC03805I 0:0401 QETH: tun0: Register guest IP address 192.168.1.115
----
12:52:19.818 00000788 HHC03997I 0:0401 QETH: tun0: not using MAC address 02:00:5e:a3:be:84
12:52:19.818 00000788 HHC03997I 0:0401 QETH: tun0: not using IP address 192.168.1.115
12:52:19.818 00000788 HHC03997I 0:0401 QETH: tun0: not using subnet mask 255.255.0.0
12:52:19.818 00000788 HHC03997I 0:0401 QETH: tun0: not using MTU 1500
--
12:52:49.462 00000788 HHC00822S PROCESSOR CP01 APPEARS TO BE HUNG!
12:52:49.462 00000788 HHC00007I Previous message from function 'watchdog_thread' at impl.c(536)
12:52:49.462 00000788 HHC00822S PROCESSOR CP03 APPEARS TO BE HUNG!
12:52:49.462 00000788 HHC00007I Previous message from function 'watchdog_thread' at impl.c(536)
12:52:49.462 00000788 HHC00822S PROCESSOR IP06 APPEARS TO BE HUNG!
12:52:49.462 00000788 HHC00007I Previous message from function 'watchdog_thread' at impl.c(536)
12:52:49.462 00000788 HHC00822S PROCESSOR IP07 APPEARS TO BE HUNG!
12:52:49.462 00000788 HHC00007I Previous message from function 'watchdog_thread' at impl.c(536)
12:52:49.462 00000788 HHC00823S You have 45 seconds to attach a debugger before crash dump will be taken!
12:52:49.462 00000788 HHC00007I Previous message from function 'watchdog_thread' at impl.c(554)
---
12:53:35.492 00000788                       ***************
12:53:35.492 00000788                       *    OOPS!    *
12:53:35.492 00000788                       ***************
12:53:35.492 00000788                     Hercules has crashed!
12:53:35.492 00000788 (you may or may not need to press ENTER if no 'oops!' dialog-box appears)
12:53:40.404 00000788 Creating crash dump "C:\hercules4.4.1\Hercules.dmp"...
12:53:40.404 00000788 Please wait; this may take a few minutes...
12:53:40.404 00000788 (another message will appear when the dump is complete)
12:53:44.470 00000788 Dump "C:\hercules4.4.1\Hercules.dmp" created.
12:53:44.470 00000788 Please forward the dump to the Hercules team for analysis.
```

**Comment by Fish-Git** [comment]
A couple of things right off the bat:

1. I would prefer seeing a _**complete**_ Hercules configuration file rather than just a single device statement. There could be other valuable information in there that might help explain things.

2. I would prefer to see a _**complete**_ Hercules log file (from beginning to end) rather than the small extract your provided. There could be valuable information in the log that might help explain what's going on (or what happened).

3. You said "I also ran TTTest64.exe successfully", but simply starting TTTest64 and then not doing anything is not enough. After starting TTTest64 you need to actually perform a test! More specifically, a "Multi-directional Ping Test", as explained in the CTCI-WIN documentation (Help file). You also failed to show us your TTTest64 test output too. It contains valuable information about your host networking setup.

4. On your Hercules device statement, you specified `netmask 255.255.0.0`, but in your z/OS TCP config you specified `255.255.255.0`.

Most of these things are mentioned in our "**[SUBMITTING PROBLEM REPORTS](https://github.com/SDL-Hercules-390/hyperion/issues/373)**" document. Please review it and then submit all needed additional information. Thanks.

In the mean time I'll take a peek at your dump to see whether anything jumps out at me.

**Comment by Fish-Git** [comment]
Dump was worthless.

I need to see the _**complete**_ Hercules log file, configuration file, and TTTest64 output.

Explaining in more detail what you were doing (or trying to do) when the problem occurred might help too.

p.s. Did you disable Unconstrained Transactions (i.e. run TXOFF) before doing whatever it was you were attempting to do, like you did **[previously](https://github.com/SDL-Hercules-390/hyperion/issues/339#issuecomment-1096712092)**?

**Comment by cfdonatucci** [comment]
Hi, thank you for your answer. The netmask was fixed and everything redone. It failed again. 
I attached the entire log, second dump, TTTest64 report and entire config file. I also ran TXOFF BEFORE doing whatever I do, the  only thing not clear to me is when TXON should be executed. I'm testing with OSA because it failed with CTCI as well. 
In general hercules works pretty fine. The issue occurs when I want to access pc applications to exploit mainframe facilities. 

Activities. 
. TXOFF is executed after iPL.
. Logon using Ip session with 192.168.1.115 port 23 works.
. CICS explorer using CMCI connection works completely.

. Rsed against 192.168.1.115 port 4035 stablishes session but then something crashes and hercules abends. 

. Connection from zosexplorer to zCEE server can be stablished. i can see services and apis.
. Postman REST services against 192.168.1.115 using tcpservice/ipcon on cics works.
. services/apis deployment crashes the same as RSED connection.

I'm using all products included in hercules, but for example I've seen that winpcap is not supported and npcap is recommended. Don't know if that may be related.

Let me know anything else you need.
Regards
Carlos

**Comment by cfdonatucci** [comment]
[Documents.zip](https://github.com/SDL-Hercules-390/hyperion/files/8643411/Documents.zip)
[Hercules2.zip](https://github.com/SDL-Hercules-390/hyperion/files/8643414/Hercules2.zip)

**Comment by Fish-Git** [comment]
I'm going to need some help reproducing this. I have almost zero z/OS skills.

I do have z/OS 2.4B, and it seems to IPL and run just fine, but for all of my z/OS Hercules IPL tests, I always use [**loadparm**](https://www.ibm.com/docs/en/zdt/12.0.0?topic=reference-loadparms-options) `WSM` (WS = CLPA and Warm start of JES2. Base z/OS system functions i.e. no CICS, DB2, IMS, etc. M=Verbose IPL Messages), not `CI`.

When I just now tried IPLing my system (which I believe is an ADCD system) using [loadparm **`CIM`**](https://www.ibm.com/docs/en/zdt/12.0.0?topic=reference-loadparms-options) (CI = CLPA and Warm start of JES2. Loads  CICS 5.3 and 5.2 libraries. Starts CICS 5.3, z/OSMF, and IBM Developer for z Systems. M=Verbose IPL Messages), it's asking me:
```
    IGGN505A SPECIFY UNIT FOR DFH540.CICS.SDFHLPA ON B4C541 OR CANCEL
```
which I don't know how to respond to. &nbsp; `:(`

I also don't know what `cicsexpl` is, nor that a `zosexpl` session with RSED is either. &nbsp; `:(`

My system is configured to use OSA and is configured almost identically as yours.

Can you explain to me (in simple terms please! I'm not a z/OS person!) what I need to do to reproduce your problem?

Thanks.
___
p.s. I suspect something might be wrong with your local network configuration. Your TTTest64 report does not look right. Your first ping of www.linux.org using a tun, reported a ping response time of "time=16ms", which is quite reasonable. But then you closed your tun interface and tried the same thing again using a tap interface instead, and this time all of your ping responses were all "time=<1ms"!! I am seriously doubting you can ping www.linux.org from your Windows system in _less_ than 1ms!!

Can you provide some more details regarding your local networking? Thanks.

**Comment by cfdonatucci** [comment]
Hi, Regarding the message is just because the CICS 54 DFHLPA maybe in the lpalist but the lib doesn't exist. In such cases you can reply 0,cancel. All consoles messages at that stage must be replied with 0,something. Sorry if this it obvious. zosexplorer is an IBM eclipse application working as a framework for other applications inluded as plugins, like cics explorer, git interfaces, zos connect enterprise edition, zosexplorer, dbb for zDevOps and others. I start my ADCD zos 2.4 with CI as well. 
To reproduce my problem you could use zosexplorer. You need two started tasks, jmon and rsed which I believe are started by default and define a connection to default port 4035 eclipse zos explorer.

With my previous windows I used to have a fixed IP adress 192.168.1.110 with a CTCI definition. Unfortunately I didn't keep record of that configuration. So I tried to do the same and now I have this problem. This worked in my other windows. 
So I switched to OSA definition hoping it'd help, but it didn't. 

I erased the ARP chahe and reexecuted TTTest64 Ping Test.
My local networking has DHCP, no other configuration.
[report.zip](https://github.com/SDL-Hercules-390/hyperion/files/8654013/report.zip)

**Comment by Fish-Git** [comment]
> Regarding the message, it is just because the CICS 54 DFHLPA maybe in the lpalist, but the lib doesn't exist. In such cases you can reply `0,cancel`.

Thanks. That seems to have worked. I had to reply to it twice though. After the first reply, the same(?) message appeared again a minute or so later. After I replied to the second message, the system finally finished IPLing and is now running normally(?).

> To reproduce my problem you could use zosexplorer.

Which I know nothing about. &nbsp;`:(`

> You need two started tasks: jmon and rsed...

Which I don't know how to do. &nbsp;`:(`

> ...which I believe are started by default...

Good!! &nbsp; `:)`

> ...and define a connection to the default eclipse zos explorer port, port 4035.

1. How can I tell whether `zosexplorer` (i.e. jmon and rsed??) is finished and ready for work? After IPLing my system, all processors are running at nearly 100%. How do I know when everything is ready for me to begin my testing? How long (approximately) do I have to wait before trying my test?

2. What do I do with port 4035? Do I connect a terminal to it? What type of terminal? 3270? Or simple command line?

3. Once connected(?), _THEN_ what do I do? Do I simply enter some type of command? What command do I enter? What do I need to try doing to reproduce the crash?

Sorry for the stupid questions, but as I explained, I know almost _nothing_ about z/OS! &nbsp;`:(`

Thanks.

**Comment by Fish-Git** [comment]
> 1. How can I tell whether `zosexplorer` (i.e. jmon and rsed??) is finished and ready for work?

FYI: When I press PF10 on my master console (to issue the `D A,L` command), I _**do**_ see both `JMON` and `RSED` in the list of active tasks. Does that mean they're both ready? Does that mean I can do my test? (whatever that test is! I don't even know what I'm supposed to be doing!). Thanks.

**Comment by cfdonatucci** [comment]
one question, do you have a TSO session or just the console?

**Comment by Fish-Git** [comment]
> one question, do you have a TSO session or just the console?

TSO session too. I should tell you I DO know how to do a little teensy tiny bit. I know how to logon. I know how to use ISPF 6 to issue 'ping' and other commands. I know how to browse dataset/edit dataset members, submit jobs (although I do _NOT_ know JCL!), look at printouts, etc. And I know how to cleanly shut the system down. But nothing beyond that,

**Comment by cfdonatucci** [comment]
ok, I can guide you to download zos explorer from IBM site, how to configure a connection to Mainframe and how it's invoked if you are willing to do it. I'll have to take a series of screenshots and send them to you. Please confirm if you want. additionally, do you have TXOFF/ON in your zos?

**Comment by Fish-Git** [comment]
> ...if you are willing to do it.

Yes, certainly!

> I'll have to take a series of screenshots and send them to you.

That's fine. My email address is fish at either softdevlabs.com or infidels.org.

> additionally, do you have TXOFF/ON in your zos?

Not yet, no. But I do have a copy of Jürgen's TXONOFF job stream, so as far as I know all I have to do is run it, and then it'll be on my system. Yes? And then all I need to do is enter the command 'txoff' or 'txon' from ISPF 6 to disable or enable unconstrained transactions, yes?

**Comment by cfdonatucci** [comment]
yes to both... i'm writing the instructions. I'll send them as soon as i can. Thank you very much.

**Comment by Fish-Git** [comment]
CORRECTION: I just found my notes regarding TXONOFF:
```
--------------------------------------------------------------------------------

                    Disable/Enable UNconstrained transactions


Upload Jürgen's TXONOFF to IBMUSER adcd.lib.jcl and run it.

(Be careful during upload! The job stream is CASE SENSITIVE!)

Then to activate, either re-IPL, or issue console command "F LLA,REFRESH"

Then to switch unconstrained transactions OFF/ON
(FOR ALL SUBSEQUENTLY STARTED jobs!), simply do:

  "S TXOFF" at the console   (or "TXOFF" from a shell prompt)
  "S TXON"  at the console   (or "TXON"  from a shell prompt)

Example:

  alias cls=clear

  export PATH=$PATH:$JAVA_HOME/bin

  TXOFF

  time java com.ibm.jvm.format.TraceFormat test.trc
  time java com.ibm.jvm.format.TraceFormat test.trc > test.trc.fmt 2>&1

--------------------------------------------------------------------------------
```

Is that correct?

**Comment by Fish-Git** [comment]
> i'm writing the instructions. I'll send them as soon as i can.

Thanks! Standing by...

**Comment by Fish-Git** [comment]
FYI: TXONOFF is now on my system, and entering `s txoff` from the master console resulted in:
```
   - 17.24.51           s txoff
   - 17.24.51           IRR812I PROFILE * (G) IN THE STARTED CLASS WAS USED
   -         TO START TXOFF WITH JOBNAME TXOFF.
   - 17.24.51 STC00442  $HASP373 TXOFF    STARTED
   - 17.24.52 STC00442  IEF404I TXOFF - ENDED - TIME=17.24.52
```

So I think we're good to go.

Standing by for further instructions via email...

**Comment by cfdonatucci** [comment]
mail sent, let me know if you got it.

**Comment by Fish-Git** [comment]
Nope. Not yet. &nbsp;`:(`

What email address did you send it to?

**Comment by cfdonatucci** [comment]
fish@softdevlabs.com

**Comment by cfdonatucci** [comment]
Hi I'll attach the file here. Regards.
[zOSexpl.docx](https://github.com/SDL-Hercules-390/hyperion/files/8662296/zOSexpl.docx)

**Comment by cfdonatucci** [comment]
Hi I did two things:
Started with CTCI support with debug. I reproduced the error but saw nothing related to CTCI.
Started with OSA support with debug. When I tried to connect to RSED I got this. Entire log attached.
I hope this helps.
[Hercules Logfile.135.zip](https://github.com/SDL-Hercules-390/hyperion/files/8662813/Hercules.Logfile.135.zip)


12:07:23.003 0000302C HHC03991D 0:0401 QETH: RRH_TYPE_ULP: PUK_TYPE_DISABLE (ULP_DISABLE): Request
12:07:23.003 0000302C HHC03981D 0:0401 QETH: TH : +0000< 00E00000 0000001B 00000014 00000055  ...............U  .\..............
12:07:23.003 0000302C HHC03981D 0:0401 QETH: TH : +0010< 10000001                             ....              ....
12:07:23.003 0000302C HHC03981D 0:0401 QETH: RRH: +0000< 00000000 417E0001 00000004 00000003  ....A~..........  .....=..........
12:07:23.003 0000302C HHC03981D 0:0401 QETH: RRH: +0010< 00240015 00001505 D8C5E3F3 00000000  .$..............  ........QET3....
12:07:23.003 0000302C HHC03981D 0:0401 QETH: RRH: +0020< 00000000                             ....              ....
12:07:23.003 0000302C HHC03981D 0:0401 QETH: PH : +0000< 01000015 00000040                    .......@          .......
12:07:23.003 0000302C HHC03981D 0:0401 QETH: PUK: +0000< 000C4103 00090000 00000000           ..A.........      ............
12:07:23.003 0000302C HHC03981D 0:0401 QETH: PUS: +0000< 00090403 05000101 16                 .........         .........
12:07:23.003 0000302C HHC03997I 0:0401 QETH: tun0: not using MAC address 02:00:5e:a3:be:84
12:07:23.003 0000302C HHC03997I 0:0401 QETH: tun0: not using IP address 192.168.1.115
12:07:23.003 0000302C HHC03997I 0:0401 QETH: tun0: not using subnet mask 255.255.255.0
12:07:23.003 0000302C HHC03997I 0:0401 QETH: tun0: not using MTU 1500
12:07:23.012 0000302C HHC03991D 0:0402 QETH: Halting data device

**Comment by Fish-Git** [comment]
> > What email address did you send it to?
> fish@softdevlabs.com

Weird. I never receive it.

> Hi, I'll attach the file here. Regards.
> 
> * [zOSexpl.docx](https://github.com/SDL-Hercules-390/hyperion/files/8662296/zOSexpl.docx)

Thanks. I downloaded it and tried it it yesterday after making the following changes:

1. I ran IBM Explorer for z/OS from another system on my local network: a Windows 7 x64 VMware virtual machine, but the fact that it was a virtual machine shouldn't make any difference. As far as my Windows 7 host was concerned, it was another system on the local network).

2. 1. Added a new Inbound rule to my Windows Firewall to let just TCP port 4035 through. That didn't seems to work, so I changed it to Protocol = Any instead, which of course removed the port number restriction, effectively disabling my firewall entirely (i.e. letting anyone connect to anything from anywhere), and that did work:

![firewall1](https://user-images.githubusercontent.com/1486786/167934283-321c7d1f-0c6a-4559-9741-0da73d5840f0.jpg)
![firewall2](https://user-images.githubusercontent.com/1486786/167934373-fbf0f445-f52b-4bbf-add2-eea1ed092ae6.jpg)
![firewall3](https://user-images.githubusercontent.com/1486786/167934393-fe9d4679-a01f-46e6-8708-cc3ec24e02e5.jpg)

As you can see below, I was able to connect and things worked just fine (although I didn't know what the heck I was doing! I'm not familiar with IBM Explorer for z/OS!):

![zosexplorer](https://user-images.githubusercontent.com/1486786/167934902-6e65ff74-9999-41f7-85d6-2de4ea257797.jpg)

I was however able to view files and printouts! Pretty cool!

One thing I did notice was that sometimes the connection would fail on the first attempt. But if I tried again it would work the second time around. I'm not sure what that means, if anything.

I also do not use DHCP. I have all of my systems are hard coded with their own uniquely assigned IP addresses, so it wasn't exactly a fair test. My initial goal wasn't to try and exactly reproduce your problem, but rather just to see if I could get it to work, and I succeeded in that endeavor.

I also have Checksum Offloading overridden in CTCI-WIN too:

    Checksum offloading = OVERRIDDEN

(Refer to "Disable CTCI-WIN's default Checksum Offload behavior" in the "Common Problems" section of the CTCI-WIN Help file)

And finally, I do not have IPv6 enabled on my adapter either (whereas you do). Despite all the hype, I've personally never found much use for IPv6. If you have a lot of internet devices maybe you have a need, I don't know. But for me, living without IPv6 is not a problem.

I would also not that at no time during my initial attempts (when my connection attempts would fail and RSED would crash (due to a TXF restricted instruction failure), at no time did my system crash. Hercules remained up and running just fine.

If I get time I will _MAYBE_ try to configure my system to use DHCP and also try running IBM Explorer for z/OS on the same system that Hercules is running on, just to see whether that makes any difference or not. I'm doubting it will, but it might be worth a shot.

Personally I think your local Windows network is borked. Your second TTTest64 report is still showing "time=<1ms" for your pings to <span>www.linux.org</span>, which is virtually impossible.

Some things to try/check:

* Check your router to make sure 192.168.1.115 (i.e. your z/OS guest's IP address) is not within its DHCP range. If it is, try using a different IP address for your z/OS guest to see if that works any better.

* Try defining your QETH (OSA) device in your Hercules configuration as a normal non-DHCP device:

    0400.3  OSA     chpid F0 iface 192.168.1.37 ipaddr 192.168.1.115 netmask 255.255.255.0

* You may or may not want to disable IPv6 on your adapter.

* Try uninstalling and reinstalling (and rebooting in between and afterwards?) CTCI-WIN. The fact that your pings to linux.org and still getting <1ms responses tells me there is still something very wrong with your networking setup. I would personally NOT PROCEED with any Hercules testing until you can _first_ get CTCI-WIN to get valid responses to its pings to linux.org. (Does the same phenomenon occur when you try pinging someone else? Such as <span>www.softdevlabs.com</span>?)

That's all for now.

I'll continue trying to reproduce your crash but so far I haven't had any luck.

**Comment by cfdonatucci** [comment]
Hi it's great you made it work. I'll take a look at your info as soon as I have time as I'm in between jobs. I erally appreciate your help. Take care.

**Comment by cfdonatucci** [comment]
Is it possible for you to send a couple of screenshots of your fixed IP definitions on windows? tks

**Comment by Fish-Git** [comment]
Here you go:

![image](https://user-images.githubusercontent.com/1486786/167996158-5a034339-d8d1-4c87-9b86-92b6cd9c10f1.png)

![image](https://user-images.githubusercontent.com/1486786/167996077-d046571c-335a-4da8-9bd3-165b864abe67.png)

![image](https://user-images.githubusercontent.com/1486786/167996411-1972abe8-a9aa-4b25-a52f-3355cf28d809.png)

![image](https://user-images.githubusercontent.com/1486786/167996509-7b4a0e10-eeff-4331-b875-d51bfbdb34b5.png)

**Comment by Fish-Git** [comment]
What happens when you try to "ping <span>www.linux.org</span>" from z/OS? (i.e. ISPF function 6) or any other IP address? Does "HOMETEST" complete successfully? Did you configure your z/OS "NSINTERADDR" DNS server values in member ADCD.Z24B.TCPPARMS(TCPDATA)?

I still think it's a problem with your Windows host's networking configuration. Since what started this whole mess was your having to reinstall your Windows 10 "after a problem" (what was the problem by the way??), Windows may have installed a default/generic device driver for your networking adapter during the install. Have you checked with your manufacturer (Realtek?) to see if there's a newer version?

Yes, I'm grasping at straws here! I admit it. But your TTTest64 ping test results keeps setting off alarm bells for me!

**Comment by Fish-Git** [comment]
Trying to launch IBM Explorer for z/OS from my Windows 7 x64 host system, I'm getting the following error dialog:

    ---------------------------
    Zosexplorer
    ---------------------------
    Java was started but returned exit code=13
    -Dorg.eclipse.swt.accessibility.UseIA2=false
    -Djava.util.Arrays.useLegacyMergeSort=true
    -XX:MaxPermSize=256m
    -Djava.class.path=C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\\plugins/org.eclipse.equinox.launcher_1.5.0.v20180512-1130.jar
    -os win32
    -ws win32
    -arch x86_64
    -showsplash
    -launcher C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\zosexplorer.exe
    -name Zosexplorer
    --launcher.library C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\\plugins/org.eclipse.equinox.launcher.win32.win32.x86_64_1.1.700.v20180518-1200\eclipse_1705.dll
    -startup C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\\plugins/org.eclipse.equinox.launcher_1.5.0.v20180512-1130.jar
    --launcher.overrideVmargs
    -showlocation
    -pluginCustomization plugin_customization.ini
    -vm C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\jre\bin\j9vm\jvm.dll
    -vmargs
    -Dorg.eclipse.swt.accessibility.UseIA2=false
    -Djava.util.Arrays.useLegacyMergeSort=true
    -XX:MaxPermSize=256m
    -Djava.class.path=C:\Users\Fish\Downloads\#489\IBM Explorer for zOS\\plugins/org.eclipse.equinox.launcher_1.5.0.v20180512-1130.jar 
    ---------------------------
    OK   
    ---------------------------

According to "https://stackoverflow.com/questions/11461607/cant-start-eclipse-java-was-started-but-returned-exit-code-13" it's because I don't have a 64-bit version of JDK installed.

I've had MANY/MUCH PROBLEMS with Java in the past on my system, so I am NOT going to try installing java to try to fix the problem. Sorry! &nbsp;`:(`

Having a stable host system for Hercules development is of paramount importance to me and I don't trust Oracle at all. Every damn time I try to do something with java it invariably ALWAYS cause me MUCH GRIEF, causing me to spend a LOT of time and effort straigtening out (fixing/undoing) the damage that Java/Oracle has done to my system!

So it looks like I'm not going to be able to reproduce your environment. &nbsp;`:(`

We'll just have to try and figure out what's wrong with your Windows 10 networking.

**Comment by Fish-Git** [comment]
Additional info:

```
C:\Program Files (x86)\MegaRAID Storage Manager\JRE\bin> java -version
java version "1.8.0-ea"
Java(TM) SE Runtime Environment (build 1.8.0-ea-b88)
Java HotSpot(TM) Client VM (build 25.0-b30, mixed mode)
```

I'm not sure what "mixed mode" means, but the fact that it's in the "Program Files (x86)" directory tells me it's a 32-bit version of java.

I suppose I _could_ download the 32-bit version of IBM Explorer for z/OS and try that. That might work. Let me think about that...

**Comment by Fish-Git** [comment]
> I suppose I _could_ download the 32-bit version of IBM Explorer for z/OS and try that. That might work. Let me think about that...

The 32-bit version doesn't work for me either. It it gets the same error:
```
    Java was started but returned exit code=13
```
So we're just going to have to debug this issue on _your_ system instead, since I'm unable to reproduce it on mine.&nbsp;`:(`

**Comment by cfdonatucci** [comment]
Yes, I think it's better because there is a lot of stuff to consider. Anyway, I was running a lot of tests trying to use a fixed IP but at some point nothing worked. So I uninstall everything, I verified that almost all entries in the registry were gone, reinstall again and trying to make it work with DHCP. The range of IPs of my router goes from 192.168.1.30 to 192.168.1.63. so IPs around 100 are not used. I defined two devices in ZOS, one CTCI and one OSA. and both were started ok. I ran all tests with the CTCI one using IP 192.168.1.112.

Once zos started I did this.

.1 HomeTEST
EZA0619I Running IBM MVS TCP/IP CS V2R4 TCP/IP Configuration Tester         
EZA0621I The FTP configuration parameter file used will be "TCPIP.FTP.DATA".  
EZA0602I TCP Host Name is: S0W1                                               
EZA0605I Using Host Tables to Resolve S0W1                                           
EZA0611I The following IP addresses correspond to TCP Host Name: S0W1                
EZA0612I 192.168.1.112                                                                                        
EZA0614I The following IP addresses are the HOME IP addresses defined in PROFILE.TCPIP:   
EZA0615I 10.1.10.1                                                                   
EZA0615I 192.168.1.112                                                               
EZA0615I 192.168.1.115                                                               
EZA0615I 10.1.10.1                                                                   
EZA0615I 127.0.0.1                                                                  
EZA0618I All IP addresses for S0W1 are in the HOME list!                             
EZA0622I Hometest was successful - all Tests Passed! 

.2 nestat HOME
EZZ2350I MVS TCP/IP NETSTAT CS V2R4       TCPIP Name: TCPIP           16:53:22 
EZZ2700I Home address list:                                                    
EZZ2701I Address          Link             Flg                                 
EZZ2702I -------          ----             ---                                 
EZZ2703I 192.168.1.112    ETH1             P                                   
EZZ2703I 192.168.1.115    OSA1                                                 
EZZ2703I 10.1.10.1        EZASAMEMVS                                           
EZZ2703I 127.0.0.1        LOOPBACK                                             
EZZ2704I Address          Interface        Flg                                 
EZZ2704I -------          ---------        ---                                 
EZZ2703I 10.1.10.1        EZAZCX 

.3 PINGS
Pinging 192.168.1.112 with 32 bytes of data:
Reply from 192.168.1.112: bytes=32 time=2ms TTL=64
Reply from 192.168.1.112: bytes=32 time=1ms TTL=64
Reply from 192.168.1.112: bytes=32 time=2ms TTL=64
Reply from 192.168.1.112: bytes=32 time=1ms TTL=64

Ping statistics for 192.168.1.112:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 1ms, Maximum = 2ms, Average = 1ms

C:\Users\Carlos>ping 192.168.1.115

Pinging 192.168.1.115 with 32 bytes of data:
Reply from 192.168.1.115: bytes=32 time<1ms TTL=64
Reply from 192.168.1.115: bytes=32 time=3ms TTL=64
Reply from 192.168.1.115: bytes=32 time=1ms TTL=64
Reply from 192.168.1.115: bytes=32 time=2ms TTL=64

Ping statistics for 192.168.1.115:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 3ms, Average = 1ms

These PINGS were executed using TSO option 6. 
CS V2R4: Pinging host linux.org (172.67.148.63)   
Ping #1 response took 0.006 seconds.              

CS V2R4: Pinging host github.com (140.82.112.4)    
Ping #1 response took 0.170 seconds. 

.4 I was able to start a TSO session using CTCI connection

.5 TXOFF executed and RSED started

.6 At this point I attempeted a connection from zos explorer.
Connection is stablished as you can see, but at this point the adapter failed. This failure can be consistently reproduced.

Options: CONN TCP TCPIP STACK TITLES ( CLI RSED*                              
EZZ2350I MVS TCP/IP NETSTAT CS V2R4       TCPIP Name: TCPIP           17:01:39
EZZ2585I User Id  Conn     Local Socket           Foreign Socket         State
EZZ2586I -------  ----     ------------           --------------         -----
EZZ2587I RSED     00000042 0.0.0.0..4035          0.0.0.0..0             Liste
EZZ2587I RSED1    0000006D 192.168.1.112..9308    192.168.1.36..50679    Estab  

and this error is issued 
11.01.15 STC09959  +FEK115E write() failed. reason=(EDC5140I Broken pipe.)  
EDC5140I Broken pipe.
Explanation: A write was attempted on a pipe or FIFO for which there was no process to read the data. This message is equivalent to the POSIX.1 EPIPE errno.
System action: The request fails. The application continues to run.
Programmer response: Refer to z/OS XL C/C++ Runtime Library Reference for the function being attempted for the specific reason for failure.

.7 Hercules and ZOS stopped and started again. I'll try with ZOS Connect.
Using zCEE I could acquire a connection as well. I could do some tasks with services and APIS. 

Options: CONN TCP TCPIP STACK TITLES ( CLI RSED* ZC*          
EZZ2350I MVS TCP/IP NETSTAT CS V2R4       TCPIP Name: TCPIP           18:09:29
EZZ2585I User Id  Conn     Local Socket           Foreign Socket         State
EZZ2586I -------  ----     ------------           --------------         -----
EZZ2587I ZCEESRV1 00000066 0.0.0.0..9001          0.0.0.0..0             Liste
EZZ2587I ZCEESRV1 00000065 0.0.0.0..9002          0.0.0.0..0             Liste
EZZ2587I ZCEESRV1 00000064 0.0.0.0..9000          0.0.0.0..0             Liste
EZZ2587I ZCEESRV1 00000067 0.0.0.0..9003          0.0.0.0..0             Liste
EZZ2587I ZCEESRV1 000000AC 192.168.1.112..9002    192.168.1.36..53681    Estab
EZZ2587I ZCEESRV1 00000063 0.0.0.0..9004          0.0.0.0..0             Liste
EZZ2587I ZCEESRV1 00000047 127.0.0.1..1025        0.0.0.0..0             Liste

.8 when I tried to deploy a service using that connection the adapter crashed again.
When Hercules is stopped it hungs. 
18:17:52.348 0000180C HHC00417I 0:0AA8 CKD file d:/ZOS240/dasd/S1C521: cache hits 460, misses 152, waits 0
18:18:11.513 * "Hercules" forcibly terminated by user request
18:18:11.513 Hercules: kill 0000180C
18:18:11.513 kill 0x0000180C  (Hercules)

Summary
It's a strange failure, because the connections are stablished but something makes then fail. I had to reinstall windows because I upgraded motherboard, cpu and memory. Hercules ran very slowly in my other hardware, now it doesn't, too bad.  I also installed another windows, from other iso file.

**Comment by Fish-Git** [comment]
> and this error is issued:
> 
> ```
> 11.01.15 STC09959  +FEK115E write() failed. reason=(EDC5140I Broken pipe.)
> ```

I had this same error before updating my Windows Firewall. The clue is:



> _EDC5140I Broken pipe._
> _**Explanation:** A write was attempted on a pipe or FIFO for which there was no process to read the data. This message is equivalent to the POSIX.1 EPIPE errno._

Try temporarily disabling your Windows Firewall, or adding a rule **[like I did further above](https://github.com/SDL-Hercules-390/hyperion/issues/489#issuecomment-1124281551=)**. _(And don't forget to disable the rule afterwards when you're no longer using it or no longer need it, and/or re-enable the Windows Firewall again if you chose to completely disable it, so that you're network security isn't left in an exposed state! This is just a temporary test after all!)_

**Comment by Fish-Git** [comment]
P.S. Also, if you haven't done so already, you should add a "Ping" rule to your Windows Firewall as well, as explained in the "Add a "Ping" rule to Windows Firewall" topic of the "Common Problems" chapter of the CTCI-WIN Help file.

**Comment by Fish-Git** [comment]
You might also need to do a network trace (e.g. [Wireshark](https://www.wireshark.org/)) to find out what's actually going on. That should let us know whether the packets are actually getting sent or not. If they are but the recipient isn't receiving them, it's more than likely the firewall.

You might need to add a custom (specific) Windows Firewall rule to let through all packets from z/OS (i.e. from IP addresses 192.168.1.112 and 1.115).

**Comment by cfdonatucci** [comment]
Hi
I did some additional test, with firewall completely disabled and with the rules, which I'm assuming i defined properly. I't didn't work. So I installed Wireshark and run a test capturing 192.168.1.112. First I login into an IP terminal ok, and saw the trace. Then I logoff. That part is not in the attached file. Then I started a session from zOS explorer using Cicsplex manager and it works perfectly. Then I did the usual stuff with RSED and I got the error. In the trace youl see a TLS error. 
10255	896.578269	192.168.1.36	192.168.1.112	TLSv1.2	61	Alert (Level: Fatal, Description: Unexpected Message)
then several 
10272	897.015414	192.168.1.112	192.168.1.36	TCP	54	[TCP Dup ACK 10269#1] 4035 → 54492 [ACK] Seq=1 Ack=28 Win=131040 Len=0
and finally many 
10386	904.250068	192.168.1.112	192.168.1.36	TCP	1514	[TCP Spurious Retransmission] 9308 → 54493 [PSH, ACK] Seq=17 Ack=653 Win=130400 Len=1460.

port 54993 was the one I was connected with. 
EZZ2350I MVS TCP/IP NETSTAT CS V2R4       TCPIP Name: TCPIP           13:40:37
EZZ2585I User Id  Conn     Local Socket           Foreign Socket         State
EZZ2586I -------  ----     ------------           --------------         -----
EZZ2587I RSED     00000044 0.0.0.0..4035          0.0.0.0..0             Liste
EZZ2587I RSED9    0000005B 192.168.1.112..9308    192.168.1.36..54493    Estab

So I have to see why TLS is involved here because If I'm right I disabled it. 
I hope this helps.
[trace220513-1347.zip](https://github.com/SDL-Hercules-390/hyperion/files/8689563/trace220513-1347.zip)

**Comment by cfdonatucci** [comment]
Hi
I Installed wireshhark and reboot the PC. I was using the portable version before.
I started the 112 trace before starting hercules. At the beginning of the trace I see the same IP 112 has two macs.
Don't know how that can be possible. After that the same error occurred.
New trace. bye.

7	153.340184	02:00:5e:a8:01:73	Broadcast	ARP	42	ARP Announcement for 192.168.1.112 (duplicate use of 192.168.1.112 detected!)
8	153.340261	02:00:5e:a8:01:70	02:00:5e:a8:01:73	ARP	42	Gratuitous ARP for 192.168.1.112 (Reply)
9	158.600785	02:00:5e:a8:01:73	02:00:5e:a8:01:73	ARP	42	Gratuitous ARP for 192.168.1.112 (Reply) (duplicate use of 192.168.1.112 detected!)
10	189.193929	AskeyCom_79:c8:10	Broadcast	ARP	60	Who has 192.168.1.112? Tell 192.168.1.1
11	189.194015	02:00:5e:a8:01:70	AskeyCom_79:c8:10	ARP	42	192.168.1.112 is at 02:00:5e:a8:01:70
12	189.194112	02:00:5e:a8:01:73	AskeyCom_79:c8:10	ARP	42	192.168.1.112 is at 02:00:5e:a8:01:73
[trace220513-1604.txt](https://github.com/SDL-Hercules-390/hyperion/files/8690633/trace220513-1604.txt)

**Comment by Fish-Git** [comment]
> and ran a test capturing 192.168.1.112.

You should have captured both 192.168.1.112 as well as 192.168.1.115 too. Earlier you said:

> I defined two devices in z/OS: one CTCI and one OSA.

And we can see the following in your HOMETEST:

> ```
> EZA0614I The following IP addresses are the HOME IP addresses defined in PROFILE.TCPIP:   
> EZA0615I 10.1.10.1                                                                   
> EZA0615I 192.168.1.112                                                               
> EZA0615I 192.168.1.115
> ```

So it looks like you've defined _both_ IP addresses to your z/OS guest. I'm presuming one of them was assigned to the CTCI device and one was assigned to the OSA device.

> Then I started a session from z/OS Explorer using Cicsplex manager, and it worked perfectly!

Fantastic!

> Then I did the usual stuff with RSED...

Wait... WHAT?!

You "did the usual stuff"? What does that mean? Do you mean to started (ran) TXOFF and then started RSED afterwards? Is that what you mean? Did you cancel/kill the existing RSED beforehand? Because if you didn't, then you would end up with _TWO_ running RSED instances, which might explain the "TCP Dup ACK" and "TCP Spurious Retransmission" errors you're seeing in your Wireshark trace.

Did you disconnect from your previous session before you "did the usual stuff with RSED"? That might explain things as well.

As far as I know, you can (_**should**_) only have _ONE_ and _**only** one_ running instance of RSED (unless each instance is listening for connections on a completely different port of course). Having multiple server instances each listening for connections on the same server port is a recipe for disaster.

> Should I install Npcap?

No.

Well... _technically_... you can if you want to. But if you do, any problems you might have with CTCI-WIN and/or Hercules networking in general are your own to resolve. Using Npcap instead of WinPCap in unsupported by CTCI-WIN. It might work, or it might not. I don't know. I've never tried it and am not interested in trying it. WinPCap works fine.

_**Conclusion:**_

Based on the fact that you _were_ able to successfully establish a z/OS Explorer, it sounds to me like your Windows Firewall was the culprit all along. Which makes sense. Hercules (specifically, z/OS's RSED) was trying to communicate with your z/OS Explorer client, and Windows Firewall wasn't letting anything through. Thus the broken pipe write failures. As soon as you disabled(?) (i.e. "fixed") your Windows Firewall issue, things started working.

NOW your only problem is having two different IP addresses assigned to your z/OS guest because you have two different adapters/interfaces defined: one CTCI and one OSA.

My suggestion would be to choose one or the other and drop the other. Personally I prefer OSA myself. While I'm sure CTCI or even LCS too would both also work just fine, OSA is more modern from a z/OS point of view and thus the device/protocol that z/OS more than likely "prefers", so that's what I'd go with: OSA.

But the choice is yours.

p.s. I personally don't recall having to mess with TLS or SSL at all, so I'm not sure what you're referring to having to do? I didn't have to change/configure anything. I just started IBM Explorer for z/OS, connected, and VOILA! I was up and running.

**Comment by Fish-Git** [comment]
> At the beginning of the trace, I see the same IP .112 has two macs. Don't know how that can be possible.

It's probably because you have two networking interfaces defined in your z/OS guest (one, a CTCI device, the other, an OSA device), _and_ (I'm presuming), you've more than likely defined _both_ of them with the same IP address. You need to either get rid of one of them or else assign it a different IP address.

**Comment by Fish-Git** [comment]
> <pre>
> 11  189.194015  02:00:5e:a8:01:70  AskeyCom_79:c8:10  ARP  42  192.168.1.112 is at 02:00:5e:a8:01:70
> 12  189.194112  02:00:5e:a8:01:73  AskeyCom_79:c8:10  ARP  42  192.168.1.112 is at 02:00:5e:a8:01:73
> </pre>

As you can see your your Wireshark trace, the two conflicting MAC address are "02:00:5e:a8:01:70" and "02:00:5e:a8:01:73". Those are MAC addresses that are automatically generated by Hercules, and correspond to IP Addresses 192.168.1.112 and 192.168.1.115. (x'70' = 112 and x'73' = 115):

> _If not specified then one will be internally generated in the range 02:00:5E:80:00:00 - 02:00:5E:FF:FF:FF using the low order 23 bits of the IPv4 address. For example, if the ipv4 address is 10.1.2.3 the generated MAC address will be 02:00:5E:81:02:03._

Methinks you need to empty your ARP cache (and/or delete one or both of the entries for .112 and .115), as well as fix your z/OS guest's networkinjg device and IP address assignments.

Once you do that (along with your existing Windows Firewall fix which you've already done), things _should_ start working just fine for you.

_**AS FAR AS THE ORIGINAL HERCULES CRASH IS CONCERNED...**_

I'm going to have to presume it's simply a side effect of your Windows Firewall that unfortunately just happened by coincidence impact Hercules. What more than likely happened was whatever packets needed to be sent/received by Hercules as part of z/OS's attempt to halt its networking adapter, got "eaten" by Windows Firewall, causing Hercules to either end up waiting forever for a response to one of its requests, or, for it to wait "too long" (i.e. longer than 20 seconds) for the response.

When that happens, Hercules's "watchdog" thread (who's interval is currently hard coded at 20 seconds) kicks in and notices once of Hercules's guest processors hasn't made any progress for the past 20 seconds (indicating something is _very wrong_ somewhere (no instruction should _ever_ take longer than 20 seconds to complete!)) and so forces a crash dump.

At least that's my working theory anyway.

* Fix your z/OS guest two adapters issue.
* Fix your duplicate IP addresses issue
* Fix your Windows Firewall issue (which you've already done)

Do that, and you should be fine.

Hope that helps!

**Comment by cfdonatucci** [comment]
Very simple test.
. Firewall completely disabled.
. netsh interface ip delete arpcache executed
. Just OSA device pointing to 192.168.1.112
. zOS 2.4 IPLed. Keep in mind that RSED doesn't start with the IPL.
. OSA started
. ping 192.168.1.112 ok.
. Once IPL is finished TXOFF executed.
. RSED started.
. connection from zosexplorer -> remote system, started
. in explorer at the bottom line I see message "IZE0100I Connected to 192.168.1.112:4035, fetching children of My Home 100%" meaning the connection was stablished and it was working. 
. OSA adapter fails.         
. Hercules crashes.             

This test can be consistently reproduced.          
I don't have duplicate MACs
I don't have firewall active
I used IPs in device definition
I cleaned the ARP cache 

Documents attached
. text capture
. shark format capture
. Hercules Log
. hercules dump

[Hercules.zip](https://github.com/SDL-Hercules-390/hyperion/files/8692128/Hercules.zip)
[Documents.zip](https://github.com/SDL-Hercules-390/hyperion/files/8692121/Documents.zip)

Nice weekend

**Comment by mcisho** [comment]
192.168.1.36 is sending Ethernet frames that are too large, with the IP packet containing a length of zero, see frames 7997 and 7999 in the last trace you provided. Check the network settings on 192.168.1.36.

**Comment by cfdonatucci** [comment]
ok, what should I test? could you be more specific please?

**Comment by mcisho** [comment]
> could you be more specific please?

No, I'm afraid I can't, I don't know anything about your machine(s), or your network. All I know is what the Wireshark trace on the Hercules host showed, i.e. that 192.168.1.36 is sending Ethernet frames that appear to be unusual.

All I can suggest is that you disable any offloads, and check the MTU that is in use.

**Comment by Fish-Git** [comment]
Ian said:
> 192.168.1.36 is sending Ethernet frames that are too large, with the IP packet containing a length of zero, see frames 7997 and 7999 in the last trace you provided. Check the network settings on 192.168.1.36.

Thank you for that, Ian! I have not had a chance to download or examine Carlos's latest postings yet.  &nbsp; (I just woke up!) &nbsp; I will do so A.S.A.P., but it sounds like you may have already found the problem.

Carlos said:
> are you refering these options?

Yes. Due to the way CTCI-WIN works, since your IBM Explorer for z/OS client is running on the same system that Hercules is running on, packets to/from your IBM Explorer client and your z/OS Hercules guest are being intercepted before they reach the actual physical Windows adapter (which is where the offloading actually occurs), resulting in Hercules receiving packets larger than it can handle. (Your Windows host thinks it is communicating with another physical system somewhere out there on your local internet and so is purposely sending "Large" packets to be efficient since it believes your adapter will properly "offload" them to smaller packets.)

But because WinPCap (CTCI-WIN) intercepts them _before_ they reach the physical adapter, the offloading is not happening and Hercules ends up receiving packets much larger than it can handle. This results in malformed packets being received by your z/OS guest (which is why is keeps trying to disable its OSA adapter as part of its error recovery).

Make sure your "Large Send Offload" and "Jumbo Frame" settings are set to "Disable". This is mentioned in the "Disable an adapter's Large Send Offload (LSO) option" section of the "Common Problems" chapter of the CTCI-WIN Help file.

I will download and examine your latest tests as soon as I've had my first cup of coffee.

**Comment by Fish-Git** [comment]
Can you post your current Hercules configuration file, please? Thanks.

**Comment by cfdonatucci** [comment]
[zos24.cnf.txt](https://github.com/SDL-Hercules-390/hyperion/files/8693474/zos24.cnf.txt)

**Comment by Fish-Git** [comment]
Thank you.

I notice you have switched to using Windows adapter 192.168.1.36, whereas before you were using 192.168.1.37. Can you post another TTTest64 report, please? Thanks.

**Comment by cfdonatucci** [comment]
my IP is changing whenever I started mi pc... now is 37 again... that's why I wanted to use the MAC.

**Comment by Fish-Git** [comment]
> my IP is changing whenever I started mi pc... now is 37 again... that's why I wanted to use the MAC.

Interesting! Usually when you use DHCP, the lease is simply renewed on the IP Address that was already previously assigned, and thus should be stable. I've never heard of a DHCP server assigning a brand new IP address before. I wonder why that's happening? Who's your DHCP server? Your router/gateway? 192.168.1.1? What manufacturer/model is it? (Not important. Just curious.)

**Comment by Fish-Git** [comment]
FYI: I noticed you've added `NETDEV  D8-5E-D3-81-FE-1D` to your configuration file. Because you have, you now shouldn't need to specify any `iface` parameter on your OSA device statement. You should now be able to just use:
```
0400.3  OSA  chpid F0  ipaddr 192.168.1.112  netmask 255.255.255.0 
```
(When `iface` is not specified, it defaults to your `NETDEV` value)

**Comment by mcisho** [comment]
> my IP is changing whenever I started mi pc

Why not use a static IP address?

**Comment by cfdonatucci** [comment]
stop the printers! ... RSED worked... 
![image](https://user-images.githubusercontent.com/30753327/168440969-127b4fcc-3895-400d-af67-564869672971.png)

**Comment by cfdonatucci** [comment]
I think it was after setting  "Large Send Offload" and "Jumbo Frame" to "Disable"

**Comment by Fish-Git** [comment]
> _**HOLD THE PRESS!**_   .......   _RSED **worked!**_

Fabtastic!! &nbsp; `:-D`

> I think it was after setting "Large Send Offload" and "Jumbo Frame" to "Disable".

More than likely.

> I'm going to check other stuff now but this is great!

Yep!! &nbsp; `:))`

**Comment by Fish-Git** [comment]
Ian said:
> Why not use a static IP address?

I agree, Ian. I also prefer using statically assigned IP Addresses too. But if the user's host adapter's IP address is DHCP assigned, the IP address might not be stable. It might change (whereas the MAC address of the adapter _never_ changes).

In my experience however, even if an adapter is DHCP assigned an IP Address, the IP Address that the DHCP server chooses it usually the same one each time, since when the DHCP assigned IP Address's DHCP lease expires, the host usually just asks for it to be renewed (which is usually always granted).

For some odd reason with Carlos however **`(*)`**, [a brand new IP address is being assigned.](https://github.com/SDL-Hercules-390/hyperion/issues/489#issuecomment-1126749438), so a static IP address can't be used in his case.
____
**`(*)`** _It's probably something Microsoft has changed in Windows. It started with Windows 8, but ever since Windows 10  (and especially with Windows 11), the quality/reliability of Windows has deteriorated quite noticeably IMHO._ &nbsp; `:(`

_That's why I've personally chosen to stick with Windows 7, which IMHO is the last best Windows operating system ever produced._

**Comment by cfdonatucci** [comment]
I tried zCEE connection and worked, and now I was able to deploy services and APIS to Mainframe... bottomline all I need in my LAB is working beutifully, amazing. Needless to say that you guys are geniuses. I really appreciate so much all your help and patience. 
Take Care! Carlos

![image](https://user-images.githubusercontent.com/30753327/168442055-2d182002-7583-4215-bc71-393c4175c41d.png)

**Comment by Fish-Git** [comment]
I've changed the status (label) for this GitHub Issue to both "Bug" as well as "Invalid (user error)" too, since, even though the root problem was the result of user error, I believe there might also be a bug in Hercules's "watchdog" thread causing it to mistakenly believe a processor is hung (or still hung) when in actuality it's not actually hung, but rather might just be taking an unusually long time to perform its error recovery.

Anyway, it's something I want to look into.

**Comment by cfdonatucci** [comment]
Hi sorry i didn't aswer. I was at the office today. Just a question, to enable jumbo frame I have 4088 or 9014 bytes. I don't remember the default value... should I set 9014?

**Comment by Fish-Git** [comment]
> ... to enable jumbo frame I have 4088 or 9014 bytes. I don't remember the default value... should I set 9014?

I don't know. It probably doesn't matter. But based solely on the following information, I would probably set it to 9014:

* [Google query "jumbo frame size"](https://www.google.com/search?q=jumbo+frame+size)
* [Wikipedia article "Jumbo frame"](https://en.wikipedia.org/wiki/Jumbo_frame)

**Comment by cfdonatucci** [comment]
I set jumbo to 9014 and disabled Large Send Offload, ran txoff and started RSED. To my surprise it didn't fail.

**Comment by cfdonatucci** [comment]
But i cannot set it higher than 9014
![image](https://user-images.githubusercontent.com/30753327/168688847-c29bdb2a-39a6-47bb-9242-949ad3bada85.png)

**Comment by Fish-Git** [comment]
> ... and disabled Large Send Offload, ran txoff and started RSED. To my surprise it didn't fail.

You were supposed to ENABLE Large Send Offload for my test.

Disabling it fixes the problem (prevents the crash).

Enabling Large Send Offload (which is the normal default setting) is what causes the crash.

I need you to ENABLE it, so the problem/crash occurs.

Once you run my test, then you can disable it again so it doesn't crash.

**Comment by cfdonatucci** [comment]
That's what I did, I enabled it... when I ran the test I had Large Send Offload enabled and Jumbo set to 9014. The connection succedded and everything worked. after a while the connection crashed. I see errors in the hercules console related to QETH adapter, but hercules didn't crashed like before. I'll send the log to you.

**Comment by cfdonatucci** [comment]
Log attached
[HerculesLogfile.zip](https://github.com/SDL-Hercules-390/hyperion/files/8704303/HerculesLogfile.zip)

**Comment by Fish-Git** [comment]
> After a while the connection crashed.

What does that mean? "The _connection_ crashed"?

> ... but Hercules didn't crash like before.

Then the test was no good. We will have to try again.

I am trying to determine the cause of the original Hercules crashing, which you claimed earlier ["**could be consistently reproduced**"](https://github.com/SDL-Hercules-390/hyperion/issues/489#issuecomment-1126630386=):

> Very simple test:
> 
> * Firewall completely disabled.
> * netsh interface ip delete arpcache executed
> * Just OSA device pointing to 192.168.1.112
> * zOS 2.4 IPLed. Keep in mind that RSED doesn't start with the IPL.
> * OSA started
> * ping 192.168.1.112 ok.
> * Once IPL is finished TXOFF executed.
> * RSED started.
> * connection from zosexplorer -> "remote" system, started
> * in explorer at the bottom line I see message _"IZE0100I Connected to 192.168.1.112:4035, fetching children of My Home 100%"_, meaning the connection _was_ established and it _was_ working.
> * OSA adapter fails.
> * Hercules crashes.
> 
> _**This test can be consistently reproduced.**_
> 
> * I don't have duplicate MACs
> * I don't have firewall active
> * I used IPs in device definition
> * I cleaned the ARP cache
> 
> Documents attached:
> 
> * text capture
> * shark format capture
> * Hercules Log
> * hercules dump
> 
> Have a nice weekend.
> 
> * [Hercules.zip](https://github.com/SDL-Hercules-390/hyperion/files/8692128/Hercules.zip)
> * [Documents.zip](https://github.com/SDL-Hercules-390/hyperion/files/8692121/Documents.zip)




If Hercules does not crash, then the test is no good. I want to recreate the original conditions that was causing your Hercules to crash, but with my debugging commands active (ptt io qeth 50000, t+400, etc...).

Thanks.

**Comment by Fish-Git** [comment]
> Log attached:
> 
> * [HerculesLogfile.zip](https://github.com/SDL-Hercules-390/hyperion/files/8704303/HerculesLogfile.zip)

Not helpful. &nbsp; `:(`

**Comment by cfdonatucci** [comment]
ok, now, I set all parameters for this to work, meaning the parameters we believed resolved the problem. The connection is stablished. So I done nothing for a while. At some point the adapter started giving "22 Invalid argument" and the connection failed. I think  we should go back to the network trace. 

22:03:22.827 00000AF0 HHC00901I 0:0401 QETH: Interface tun0, type TUN opened
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using MAC address 02:00:5e:a3:be:84
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using IP address 192.168.1.112
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using subnet mask 255.255.255.0
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using MTU 1500
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using drive MAC address 96:7a:59:e5:d2:bf
22:03:22.917 00000AF0 HHC03997I 0:0401 QETH: tun0: using drive IP address fe80::967a:59ff:fee5:d2bf
22:03:22.924 00000AF0 HHC03805I 0:0401 QETH: tun0: Register guest IP address 192.168.1.112
22:03:22.925 00000AF0 HHC03805I 0:0401 QETH: tun0: Register guest IP address 10.1.10.1
22:03:34.693 00000AF0 HHC00801I Processor CP00: Operation exception code 0001 ilc 4
22:03:34.694 00000AF0 HHC02324I CP00: PSW=0704100080000000 000000001FB179D6 INST=B2AF0000     ????? ,                      
.....
23:38:53.653 00000AF0 HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument
23:38:53.653 00000AF0 HHC00007I Previous message from function 'write_packet' at qeth.c(2796)
23:38:53.663 00000AF0 HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument
23:38:53.663 00000AF0 HHC00007I Previous message from function 'write_packet' at qeth.c(2796)
23:38:54.127 00000AF0 HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument
23:38:54.127 00000AF0 HHC00007I Previous message from function 'write_packet' at qeth.c(2796)
23:38:54.923 00000AF0 HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument
23:38:54.923 00000AF0 HHC00007I Previous message from function 'write_packet' at qeth.c(2796)
23:38:55.922 00000AF0 HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument

**Comment by Fish-Git** [comment]
I would _prefer_ that we recreate [the Hercules crash that you originally reported](https://github.com/SDL-Hercules-390/hyperion/issues/489#issue-1228118977=). That is what I am most interested in: preventing Hercules from crashing.

Since we are having trouble accomplishing that however, we _might_ be able to determine a likely cause using your current testing procedure. So let's do this:

Do the same thing you did above (that caused the `22 Invalid argument` errors), and then enter the Hercules `ptt` command whenever it occurs. That will display the ptt trace table which _might_ provide a clue to what's going on:

* Start Hercules
* enter `ptt io qeth 50000`
* IPL
* `s txoff`
* `s rsed`
* connect from zosexplorer
* wait for the `HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument` errors to begin occuring
* as quickly as possible, enter `ptt` to dump the trace table (i.e. just "ptt" by itself, without any arguments)

That _should_ display all of the trace entries in the ptt trace table. _THAT_ is what I need to see.

Save the Hercules logfile and post it here.

Thanks.

**Comment by cfdonatucci** [comment]
Hi Test done. Hercules didn't crashed and didn't failed, This is the summary:

. adapter parameters changed as before.
. arp cache deleted
. Start Hercules
. enter ptt io qeth 50000
. IPL
. s txoff
. s rsed
. connect from zosexplorer

At this point everything worked perfectly. i couldn't reproduce the failure. I waited for a while and then went to lunch. when I returned the PC was sleeping. When it was awake again the connection with zosexplorer failed and the invalid argument was issued then I executed ptt and after a little while I stopped hercules.
So now I cannot reproduce the previous consistent error.
Regards
[HerculesLogfile.152.zip](https://github.com/SDL-Hercules-390/hyperion/files/8718508/HerculesLogfile.152.zip)

**Comment by Fish-Git** [comment]
> Hi. Test done. Hercules didn't crashed and didn't failed, This is the summary:

> Regards.
> 
> * [HerculesLogfile.152.zip](https://github.com/SDL-Hercules-390/hyperion/files/8718508/HerculesLogfile.152.zip)

Thanks. Unfortunately however, the test is no good. You didn't do it correctly. You entered the `ptt` command too soon: you entered it _before_ the `22 Invalid argument` error messages were issued, so the ptt results did not show anything unusual.

I need you to wait until the `22 Invalid argument` error(s) is/are issued, and _then_ issue the `ptt` command.

I've also been thinking: if you previously added a Windows Firewall rule, maybe you need to disable it to reproduce the original problem?

Please try again, being careful to issue the `ptt` command _**after**_ the `HHC00911E 0:0402 QETH: Error writing to device tun0: 22 Invalid argument` error(s) to begin occurring.

Thanks.

**Comment by cfdonatucci** [comment]
I'm going to do the test now. When I enabled the firewall rules the problem was not resolved. The problem was resolved (hold the press, remember?) when I disabled jumbo and Large Send Offload. I disabled the firewall rules. Let's see what happens.

**Comment by Fish-Git** [comment]
> When I enabled the firewall rules the problem was not resolved. The problem was resolved (hold the press, remember?) when I disabled jumbo and Large Send Offload.

What I remember is, _originally_, you did not have any special Firewall rules defined. Your Windows Firewall was enabled (which is the default for Windows).

I then recommended that you define a new Firewall rule that would allow any/all packets through (effectively disabling your firewall). I don't recall whether you actually added (defined) such a rules to your Windows Firewall or not, but I do remember that you disabled the Windows Firewall completely, since that was the easiest way to eliminate it as a culprit.

It was sometime _after_ that test (which still failed) that Ian Shorter (@mcisho) [noticed your Wireshark trace showed packets trying to be sent that were too large](https://github.com/SDL-Hercules-390/hyperion/issues/489) for CTCI-WIN (WinPCap) / Hercules to handle. It was then I recommended disabling Large Send Offload, which, yes, did indeed resolve your problem.

So what I am trying to do is to go back to the way you originally had your system configured that was causing Hercules to crash. If we can recreate your original crash again, then we can see where things might be going wrong. That is what the `ptt io qeth 50000` command should do: immediately before Hercules crashes, Hercules will automatically issue the `ptt` command automatically, so we can see what Hercules was doing.

If we cannot recreate the original crash, then we should try to recreate the "Error writing to device tun0: 22 Invalid argument" error instead (with `ptt io qeth 50000` active), and then _manually_ enter the `ptt` command _after_ the "22 Invalid argument" error(s) start to occur.

Thanks.

**Comment by cfdonatucci** [comment]
I really don't think the firewall is the issue. I ran tests with the firewall completely disabled and they failed. So far I couldn't make it fail. I'm going to run a last test.

**Comment by cfdonatucci** [comment]
I was unable to recreate the original error. To recreate "22 invalid argument" I IPLed hercules as you said. Nothing happened. So I sent the PC to sleep. After a few minutes I recalled it and the error ocurred inmediatly. after few 22 messages i executed ptt. I hope this helps. Let me know.
[Hercules Logfile.155.zip](https://github.com/SDL-Hercules-390/hyperion/files/8734667/Hercules.Logfile.155.zip)

**Comment by Fish-Git** [comment]
> I really don't think the firewall is the issue.

I do.

When you reported your original problem, the Firewall was enabled, and _because_ it was, it was not allowing any incoming connections. That was part of the original problem. Your zosexplorer client was trying to connect to RSED, but Windows wasn't allowing it because it wasn't allowing any outside packets through the firweall. It was only when we disabled the firewall that we then made some progress. Think. How was your Windows system configured when you first reported [your original problem](https://github.com/SDL-Hercules-390/hyperion/issues/489#issue-1228118977=)? That is the same configuration I would like to try to recreate.

But forget about that right now. Let's move on.

> So I sent the PC to sleep. After a few minutes I recalled it and the error occurred immediately. After few "22" messages, I executed `ptt`. I hope this helps. Let me know.
> 
> * [Hercules Logfile.155.zip](https://github.com/SDL-Hercules-390/hyperion/files/8734667/Hercules.Logfile.155.zip)

Yes, that helped. Thank you.

I would now like you to perform the same test, but this time with the following Hercules commands active:

* ptt io qeth 5000
* tt32 debug

This should provide additional (hopefully more detailed) debugging information when the "22" errors occur.

What I suspect is happening is, your network adapter (NIC) is being turned off when your PC goes to sleep, but is not being fully powered on after it wakes up. This is a known problem in Windows:

* https://docs.microsoft.com/en-us/troubleshoot/windows-client/networking/power-management-on-network-adapter

> "When deploying Windows 7 or Windows Server 2008 R2, you may want to _**disable**_ the following network adapter power management setting on some computers:"
> 
> **Allow the computer to turn off this device to save power**

_(emphasis mine)_

And:

> "Windows never turns off the network card due to inactivity. When this setting is checked (enabled), Windows puts the network card to sleep, and when it resumes, it puts it back to D0. When this setting isn't checked (disabled), Windows _**completely halts**_ the device, and on resume _**reinitializes**_ it."

_(emphasis mine)_

I also found the following web page as well:

* https://community.spiceworks.com/topic/2084628-gpo-to-uncheck-allow-the-computer-to-turn-off-this-device-to-save-power-on-nic

> "By trial and error and experimentation, I have determined that if I uncheck that box, the issue disappears."

(and, at the very bottom of the following page):

* https://supportcenter.pleora.com/s/article/Configuring-Your-Computer-and-Network-Adapters-for-Best-Performance-KBase#Configuring_Power_Management_Windows

> **Configuring Power Management (Windows Operating System Only)**
>
> "For image streaming applications that run for an extended period of time, we recommend that you change the Windows power plan settings to ensure the computer and NIC never go to sleep. If you do not change the power plan settings, the computer or NIC may go to sleep when there is no activity from the keyboard or mouse for certain amount of time and the image stream may be interrupted."

(and):

> **To configure the power settings for the NIC**
>
> 1. In the Windows Control Panel, click Network and Internet.
> 2. Click Network and Sharing Center.
> 3. In the left-hand panel, click Change adapter settings.
> 4. Right-click the NIC and then click Properties.
> 5. Click Configure.
> 6. Configure the power management settings for the NIC to ensure the computer _does **not**_ turn off the device to save power.
> 7. Click OK.
> Repeat steps 4-7 for each NIC on your computer.

_(emphasis mine)_

_**Bottom line:**_ everywhere I look, _disabling_ (unchecking) the "Allow the computer to turn off this device to save power" option fixes the problem.

But I would like to see another test like you just did, but with `tt32 debug` enabled as well. I'm hoping _maybe_ it might confirm that the reason for the "22 invalid argument" errors is because the adapter isn't fully powered on after waking up from a sleep.

And if that is true, there is nothing Hercules nor CTCI-WIN can do to fix or workaround the problem. It is just another "User Error" (i.e. not having your system configured properly).

Thanks.
___
**MORE INFO:**

![nic-properties](https://user-images.githubusercontent.com/1486786/169627463-aff4c5e6-d8f5-4ef2-ae3f-d8db8420e936.jpg)

![nic-configure](https://user-images.githubusercontent.com/1486786/169627478-44d413d5-72d6-4c7b-af85-9fd8e9a7c9c7.jpg)

![nic-power](https://user-images.githubusercontent.com/1486786/169627581-1699e55b-3e20-4d4c-87f5-7e01fcf3a2ab.jpg)

**Comment by cfdonatucci** [comment]
Hi I ran the test and got your trace.
[Hercules Logfile.157.txt.zip](https://github.com/SDL-Hercules-390/hyperion/files/8748035/Hercules.Logfile.157.txt.zip)

Afterwars I unchecked that box and restarted Hercules and ZOS. after a while I put the PC to sleep. when it restarted I got no 22 error but the adapter stopped working. Apparently the suspend introduces some problem. Anyway, I set my PC to never sleeps. So I think we are ok unless you want another test.

**Comment by Fish-Git** [comment]
> Hi I ran the test and got your trace:
> 
> * [Hercules Logfile.157.txt.zip](https://github.com/SDL-Hercules-390/hyperion/files/8748035/Hercules.Logfile.157.txt.zip)

Unfortunately, your .zip file is completely empty. It contains .... nothing. No files at all.

**Comment by cfdonatucci** [comment]
[Hercules Logfile.157.txt.zip](https://github.com/SDL-Hercules-390/hyperion/files/8748308/Hercules.Logfile.157.txt.zip)
this one?

**Comment by cfdonatucci** [comment]
this one
[HerculesLogfile157.zip](https://github.com/SDL-Hercules-390/hyperion/files/8748320/HerculesLogfile157.zip)

**Comment by Fish-Git** [comment]
> Try this one:
> 
> * [HerculesLogfile157.zip](https://github.com/SDL-Hercules-390/hyperion/files/8748320/HerculesLogfile157.zip)

Wow. I wasn't expecting that. Not very helpful at all, but not your fault of course.

It looks like Windows's re-initialization of the network adapter upon resume after suspension ends up killing Hercules's guest's virtual network adapter. I'm guessing that CTCI-WIN's WinPCap sniffing/injection hook on Windows's real physical adapter ends up being invalidated/removed/rendered inaccessible resulting in a permanent loss of connectivity between Hercules (CTCI-WIN/WinPCap) and the adapter. Most unfortunate.

So it looks like setting one's PC to never sleep (which you have already done) is the only solution available to us. There doesn't appear to be any way for us (Hercules/CTCI-WIN/WinPCap) to recover from this.

Nevertheless, since my ultimate goal for this issue was not to try to overcome that problem, but rather to try to prevent Hercules from crashing as a result of the situation, I have added some additional code to Hercules (commit 96b606adad02296f2e11a8252ea9c9d7ccc1d6f8) to try and better detect when a resume from suspend (awake from sleep or re-power-on after hibernation) has occurred in order to prevent our watchdog thread from erroneously detecting what _**it**_ believes is a malfunctioning CPU.

_Hopefully_, this new code, along with our previous attempt to address the same issue (commit 4b439ae673d776ecaa35ed0aa87bb73f940e904e), will, _together_, prevent Hercules from crashing upon resume from suspend.

_&lt;me: fingers crossed&gt;_

As I believe we have done all that we can with this issue, I am going to close it at this time. I would like to thank you, Carlos, for your incredible patience. Take care.

p.s. to Mark G. (@dasdman): I apologize for not having implemented **[your earlier suggestion](https://github.com/SDL-Hercules-390/hyperion/issues/458#issuecomment-1001245586=)** sooner. At the time, I didn't think it was necessary!

**Comment by cfdonatucci** [comment]
I'm glad this was useful. It was great for me to contribute with you guys, I really admire what you do. Let me know any way I can do anything in the future, tests or whatever. I'm not at your level but I'll try to do my best.
See you!
Carlos
